### ISSUE TITLE
Custom Roles for Specific Lambdas

### ISSUE THREAD

**Issue by ac360** [serverless/serverless#174]
Currently, the JAWS Lambda role is applied to all lambdas.  We need to allow for lambdas to use custom roles.  The way I see this being implemented now is:

* Add a `role` property to the `lambda` property in awsm.json.  
* On deploy, this custom role is added, instead of the default Lambda Role

**Comment by Nopik** [comment]
What happens to default lambda role? What if no lambda will use default and each one will use custom role - should the default one still be created?

`--role` should be provided for cli as well, I guess, during `jaws module create ...`. That command also should update `resources-cf.json` for all stages (which leads to issue I've described in #145 and more general stuff of automatic `resources-cf.json` management).

**Comment by shortjared** [comment]
I think it's okay to keep the default role in place. If you are doing so  many differing things that you need custom roles for everything, I think it is a safe bet that you would be able to update the default role itself to be a solid "default". If not, an extra default role rolling around won't kill anyone.

On the cli `--role`.

Is there any need to distinguish new role vs existing role? `jaws create module users list --role DynamoDBWrite` could easily check and see if that role already exists in `resources-cf.json`. If it does, just add it as a role reference, otherwise create it is a new role in resources with a message that the role was newly created and should be updated. 

`jaws module install`, and awsm authors need to know to add those roles to their resources awsm.json.

**Comment by Nopik** [comment]
Right, I thought about autodiscovery as well, forgot to mention it ;) Yet, after writing previous comment I've realized that situation is slightly more complex. Namely, if I create new role, I typically want to use it for something, not just having separate role for the sake of having it. I.e. the new role should have e.g. sqs use permission or access to dynamo or whatever. `--role` doesnt really cut it (it works well in referencing existing roles, though). I guess we should create module and then allow user to fill in the role details in `awsm.json`. So, maybe always creating the role would be the way to go.

**Comment by shortjared** [comment]
Correct, which is why I suggested the role should be added to the resources-cf.json and the user is prompted to update it there if it didn't exist prior. However, I did just realize if you are running multiple stages / regions... resources-cf.json gets its own json in the directory tree for each permutation. 

That is a large problem in this situation.

We need some resources sync functionality. Manually maintaining resources-cf.json is not sustainable across many regions and / or stages.

**Comment by Nopik** [comment]
Exactly, thats my point. Some resources-cf.json sync will be a great relief.

Now, how to to this? Existence of multiple awsm modules in a project makes it really tricky, as we basically want to share some resources between them. E.g. if I have 1 dynamo table and 5 lambda working on it, there is no obvious awsm.json file I could put definition to. Putting resources into somewhere into project-level place would solve this, but it will make awsm modules, well, less awesome.
Putting resource definition (dynamo table in my example) into each lambda operating on it seems even worse, crying for discrepancies between definitions. It really looks like we're missing some abstraction layer here, or some other fundamental concept.

**Comment by Nopik** [comment]
Hm. Resource-only awsm modules? With lambda modules depending on them?

**Comment by eahefnawy** [comment]
Done!

**Comment by BerndWessels** [comment]
@Nopik @eahefnawy Can you please give some details on how to create/assign a custom role to a Lambda function in serverless?
This issue seems to deal with that question but I can't see the answer or find any documentation.
We need an individual role for each lambda function with a specialized inline policy.
Thanks
Bernd
